### PR TITLE
Fix Two Bugs around ChannelManager serialization round-trip

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Based on #455 since one of the new tests builds on the tests added there. The pre-block-connected bug I found while playing with rust-lightning-bitcoinrpc, the other one I found while trying to write a test for the first.

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -3670,12 +3670,15 @@ impl<ChanSigner: ChannelKeys + Writeable> Writeable for Channel<ChanSigner> {
 		}
 		(self.pending_inbound_htlcs.len() as u64 - dropped_inbound_htlcs).write(writer)?;
 		for htlc in self.pending_inbound_htlcs.iter() {
+			if let &InboundHTLCState::RemoteAnnounced(_) = &htlc.state {
+				continue; // Drop
```
Comment (ariard) [inline lightning/src/ln/channel.rs]: nit: good to add drop rational, "as non-committed will be reannounced by peer upon reconnection"

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -3310,7 +3310,7 @@ impl<'a, R : ::std::io::Read, ChanSigner: ChannelKeys + Readable<R>, M: Deref> R
 		let mut short_to_id = HashMap::with_capacity(cmp::min(channel_count as usize, 128));
 		for _ in 0..channel_count {
 			let mut channel: Channel<ChanSigner> = ReadableArgs::read(reader, args.logger.clone())?;
-			if channel.last_block_connected != last_block_hash {
+			if channel.last_block_connected != Default::default() && channel.last_block_connected != last_block_hash {
```
Comment (ariard) [inline lightning/src/ln/channelmanager.rs]: Same from commit message, comment previous bug here where an already-created but block-not-seen channel would have its last_block_connected setup to Default

File: lightning/src/ln/functional_tests.rs [hunk 3]
```diff
@@ -413,6 +413,104 @@ fn test_1_conf_open() {
...
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	if steps & 0b1000_0000 != 0{
+		let header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
+		nodes[0].block_notifier.block_connected_checked(&header, 1, &Vec::new(), &[0; 0]);
+		nodes[1].block_notifier.block_connected_checked(&header, 1, &Vec::new(), &[0; 0]);
+	}
+
+	if steps & 0x0f == 0 { return; }
```
Comment (ariard) [inline lightning/src/ln/functional_tests.rs]: Can you comment permutation check assumptions?

**Comment by TheBlueMatt** [comment]
Added comments everywhere requested.

**Comment by ariard** [comment]
ACK bc90c80

**Pull request merged**